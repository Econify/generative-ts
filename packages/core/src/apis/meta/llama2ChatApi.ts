/* eslint-disable camelcase */
import type { ModelApi, ModelRequestOptions } from "@typeDefs";

import { FnTemplate } from "../../utils/Template";

import type { FewShotRequestOptions } from "../shared";

import { isLlamaResponse, LlamaResponse } from "./llama";

/**
 * @category Requests
 * @category Llama2
 */
export interface Llama2ChatOptions
  extends FewShotRequestOptions,
    ModelRequestOptions {
  temperature?: number;
  top_p?: number;
  max_gen_len?: number;
}

/**
 * @category Templates
 * @category Llama2
 *
 */
export const Llama2ChatTemplate = new FnTemplate(
  ({
    $prompt,
    system,
    examplePairs,
    temperature,
    top_p,
    max_gen_len,
  }: Llama2ChatOptions) => {
    const rewritten = {
      prompt: [
        `<s>[INST] `,
        ...(system ? [`<<SYS>>\n${system}\n<</SYS>>\n\n`] : []),
        ...(examplePairs
          ? examplePairs.flatMap((pair) => [
              `${pair.user} [/INST] ${pair.assistant} </s><s>[INST] `,
            ])
          : []),
        `${$prompt} [/INST]`,
      ].join(""),
    };

    const result = {
      ...rewritten,
      temperature,
      top_p,
      max_gen_len,
    };

    return JSON.stringify(result, null, 2);
  },
);

export interface Llama2ChatApi
  extends ModelApi<Llama2ChatOptions, LlamaResponse> {}

/**
 *
 * ## Reference
 * [LLama2](https://llama.meta.com/docs/model-cards-and-prompt-formats/meta-llama-2/)
 *
 * ## Providers using this API
 * - {@link createAwsBedrockModelProvider | AWS Bedrock}
 *
 * @category APIs
 * @category Llama2
 * @category Provider: AWS Bedrock
 *
 */
export const Llama2ChatApi: Llama2ChatApi = {
  requestTemplate: Llama2ChatTemplate,
  responseGuard: isLlamaResponse,
};
