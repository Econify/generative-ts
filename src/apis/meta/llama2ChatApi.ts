import type { ModelApi, ModelRequestOptions } from "../../typeDefs";

import { Template } from "../../utils/template";

import type { FewShotRequestOptions } from "../shared/fewShot";

import { isLlamaResponse, LlamaResponse } from "./llama";

const llama2ChatPrompt =
  "<s>[INST] " +
  `{% if system %}<<SYS>>\\n{{ system | safe }}\\n<</SYS>>\\n\\n{% endif %}` +
  "{% for pair in examplePairs %}{{ pair.user | safe }} [/INST] {{ pair.assistant | safe }} </s><s>[INST] {% endfor %}" +
  "{{ prompt | safe }} [/INST]";

export const templateSource =
  "{" +
  `"prompt": "${llama2ChatPrompt}"` +
  '{% if temperature %}, "temperature": {{ temperature }}{% endif %}' +
  '{% if top_p %}, "top_p": {{ top_p }}{% endif %}' +
  '{% if max_gen_len %}, "max_gen_len": {{ max_gen_len }}{% endif %}' +
  "}";

export interface Llama2ChatOptions
  extends FewShotRequestOptions,
    ModelRequestOptions {
  temperature?: number;
  top_p?: number;
  max_gen_len?: number;
}

export const Llama2Template = new Template<Llama2ChatOptions>(templateSource);

export const Llama2ChatApi: ModelApi<Llama2ChatOptions, LlamaResponse> = {
  requestTemplate: Llama2Template,
  responseGuard: isLlamaResponse,
};
