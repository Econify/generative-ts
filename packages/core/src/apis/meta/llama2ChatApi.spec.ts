import { Llama2ChatApi, Llama2ChatOptions } from "./llama2ChatApi";

function render(context: Omit<Llama2ChatOptions, "modelId">) {
  try {
    const rendered = Llama2ChatApi.requestTemplate.render({
      modelId: "mock-model-id",
      ...context,
    });
    return JSON.parse(rendered);
  } catch (error) {
    throw new Error("Invalid JSON output");
  }
}

describe("Llama2ChatApi:", () => {
  /**
   * FewShotRequestOptions (prompt, examplePairs, system):
   */

  test("prompt", () => {
    const rendered = render({
      prompt: "mock-prompt",
    });

    expect(rendered).toMatchSnapshot();
  });

  test("prompt, examplePairs", () => {
    const rendered = render({
      prompt: "mock-prompt",
      examplePairs: [
        { user: "mock-user-msg-1", assistant: "mock-assistant-msg-1" },
        { user: "mock-user-msg-2", assistant: "mock-assistant-msg-2" },
      ],
    });

    expect(rendered).toMatchSnapshot();
  });

  test("prompt, system", () => {
    const rendered = render({
      prompt: "mock-prompt",
      system: "mock-system-text",
    });

    expect(rendered).toMatchSnapshot();
  });

  test("prompt, examplePairs, system", () => {
    const rendered = render({
      prompt: "mock-prompt",
      examplePairs: [
        { user: "mock-user-msg-1", assistant: "mock-assistant-msg-1" },
      ],
      system: "mock-system-text",
    });

    expect(rendered).toMatchSnapshot();
  });

  /**
   * All options:
   */

  test("all options", () => {
    const rendered = render({
      prompt: "mock-prompt",
      examplePairs: [
        { user: "mock-user-msg-1", assistant: "mock-assistant-msg-1" },
        { user: "mock-user-msg-2", assistant: "mock-assistant-msg-2" },
        { user: "mock-user-msg-3", assistant: "mock-assistant-msg-3" },
      ],
      system: "mock-system-text",
      temperature: 0.7,
      top_p: 0.9,
      max_gen_len: 1500,
    });

    expect(rendered).toMatchSnapshot();
  });
});
