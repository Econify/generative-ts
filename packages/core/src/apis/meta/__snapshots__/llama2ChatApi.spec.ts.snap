// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Llama2ChatApi: $prompt 1`] = `
{
  "prompt": "<s>[INST] mock-prompt [/INST]",
}
`;

exports[`Llama2ChatApi: $prompt, examplePairs 1`] = `
{
  "prompt": "<s>[INST] mock-user-msg-1 [/INST] mock-assistant-msg-1 </s><s>[INST] mock-user-msg-2 [/INST] mock-assistant-msg-2 </s><s>[INST] mock-prompt [/INST]",
}
`;

exports[`Llama2ChatApi: $prompt, examplePairs, system 1`] = `
{
  "prompt": "<s>[INST] <<SYS>>
mock-system-text
<</SYS>>

mock-user-msg-1 [/INST] mock-assistant-msg-1 </s><s>[INST] mock-prompt [/INST]",
}
`;

exports[`Llama2ChatApi: $prompt, system 1`] = `
{
  "prompt": "<s>[INST] <<SYS>>
mock-system-text
<</SYS>>

mock-prompt [/INST]",
}
`;

exports[`Llama2ChatApi: all options 1`] = `
{
  "max_gen_len": 1500,
  "prompt": "<s>[INST] <<SYS>>
mock-system-text
<</SYS>>

mock-user-msg-1 [/INST] mock-assistant-msg-1 </s><s>[INST] mock-user-msg-2 [/INST] mock-assistant-msg-2 </s><s>[INST] mock-user-msg-3 [/INST] mock-assistant-msg-3 </s><s>[INST] mock-prompt [/INST]",
  "temperature": 0.7,
  "top_p": 0.9,
}
`;
